crypto/crypto.go

```diff
@@ -158,37 +159,67 @@ func FromECDSAPub(pub *ecdsa.PublicKey) []byte {
 // HexToECDSA parses a secp256k1 private key.
 func HexToECDSA(hexkey string) (*ecdsa.PrivateKey, error) {
 	b, err := hex.DecodeString(hexkey)
-	if err != nil {
-		return nil, errors.New("invalid hex string")
+	if byteErr, ok := err.(hex.InvalidByteError); ok {
+		return nil, fmt.Errorf("invalid hex character %q in private key", byte(byteErr))
+	} else if err != nil {
+		return nil, errors.New("invalid hex data for private key")
 	}
 	return ToECDSA(b)
 }
 
 // LoadECDSA loads a secp256k1 private key from the given file.
 func LoadECDSA(file string) (*ecdsa.PrivateKey, error) {
-	stat, err := os.Stat(file)
+	fd, err := os.Open(file)
 	if err != nil {
 		return nil, err
 	}
-	size := stat.Size()
-	// Allow two extra chars for possible line ending to be checked later
-	if size < 64 || size > 66 {
-		return nil, fmt.Errorf("expected 64 bytes, got %v", size)
-	}
-	buf, err := ioutil.ReadFile(file)
+	defer fd.Close()
+
+	r := bufio.NewReader(fd)
+	buf := make([]byte, 64)
+	n, err := readASCII(buf, r)
 	if err != nil {
 		return nil, err
+	} else if n != len(buf) {
+		return nil, fmt.Errorf("key file too short, want 64 hex characters")
 	}
-	buf = bytes.TrimSpace(buf)
-	if len(buf) != 64 {
-		return nil, fmt.Errorf("expected 64 bytes, got %v", len(buf))
+	if err := checkKeyFileEnd(r); err != nil {
+		return nil, err
 	}
 
-	key, err := hex.DecodeString(string(buf))
-	if err != nil {
-		return nil, err
+	return HexToECDSA(string(buf))
+}
+
+// readASCII reads into 'buf', stopping when the buffer is full or
+// when a non-printable control character is encountered.
+func readASCII(buf []byte, r *bufio.Reader) (n int, err error) {
+	for ; n < len(buf); n++ {
+		buf[n], err = r.ReadByte()
+		switch {
+		case err == io.EOF || buf[n] < '!':
```

Comment:
Isn't there a less hacky way to check a char, along the lines of Python's `buf[n] in string.printable`?

Reply:
There is no better way to do this.

Reply:
I think you could use unicode.IsPrint(rune r) https://golang.org/pkg/unicode/#IsPrint

Reply:
Could, but don't want to import "unicode" just for checking if something is ASCII.

Reply:
thats fair